src/subscription/__tests__/subscribe-test.js

```diff
@@ -0,0 +1,215 @@
+  });
+
+  const QueryType = new GraphQLObjectType({
+    name: 'Query',
+    fields: {
+      inbox: { type: InboxType },
+    }
+  });
+
+  const EmailEventType = new GraphQLObjectType({
+    name: 'EmailEvent',
+    fields: {
+      email: { type: EmailType },
+      inbox: { type: InboxType },
+    }
+  });
+
+  const SubscriptionType = new GraphQLObjectType({
+    name: 'Subscription',
+    fields: {
+      importantEmail: { type: EmailEventType },
+    }
+  });
+
+  const emailSchema = new GraphQLSchema({
+    query: QueryType,
+    subscription: SubscriptionType
+  });
+
+  it('produces a payload per subscription event', async () => {
+
+    const pubsub = new EventEmitter();
+
+    const data = {
+      inbox: {
+        emails: [
+          {
+            from: 'joe@graphql.org',
+            subject: 'Hello',
+            message: 'Hello World',
+            unread: false,
+          },
+        ],
+      },
+      importantEmail() {
+        return eventEmitterAsyncIterator(pubsub, 'importantEmail');
+      }
+    };
+
+    function sendImportantEmail(newEmail) {
+      data.inbox.emails.push(newEmail);
+      // Returns true if the event was consumed by a subscriber.
+      return pubsub.emit('importantEmail', {
+        importantEmail: {
+          email: newEmail,
+          inbox: data.inbox,
+        }
+      });
+    }
+
+    const ast = parse(`
+      subscription ($priority: Int = 0) {
+        importantEmail(priority: $priority) {
+          email {
+            from
+            subject
+          }
+          inbox {
+            unread
+            total
+          }
+        }
+      }
+    `);
+
+    // GraphQL `subscribe` has the same call signature as `execute`, but returns
+    // AsyncIterator instead of Promise.
+    const subscription = subscribe(
+      emailSchema,
+      ast,
+      data,
+      null, // context
+      { priority: 1 }
+    );
+
+    // Wait for the next subscription payload.
+    const payload = subscription.next();
+
+    // A new email arrives!
+    expect(sendImportantEmail({
+      from: 'yuzhi@graphql.org',
+      subject: 'Alright',
+      message: 'Tests are good',
+      unread: true,
+    })).to.equal(true);
+
+    // The previously waited on payload now has a value.
+    expect(await payload).to.deep.equal({
+      done: false,
+      value: {
+        data: {
+          importantEmail: {
+            email: {
+              from: 'yuzhi@graphql.org',
+              subject: 'Alright',
+            },
+            inbox: {
+              unread: 1,
+              total: 2,
+            },
+          },
+        },
+      },
+    });
+
+    // Another new email arrives, before subscription.next() is called.
+    expect(sendImportantEmail({
+      from: 'hyo@graphql.org',
+      subject: 'Tools',
+      message: 'I <3 making things',
+      unread: true,
+    })).to.equal(true);
+
+    // The next waited on payload will have a value.
+    expect(await subscription.next()).to.deep.equal({
```

Comment:
This is one way to use the AsyncIterable API, but probably not how most will use it. `next()` returns a `Promise<IteratorResult>`, so `await asyncIterator.next()` produces `IteratorResult` objects to continue to receive payloads.

A state-3 JS proposal gives us `for await` loop syntax. However to test events happening in a certain order, I've used the lower level direct API in the test.

In practice, people will use an adapter which converts AsyncIterator to some network protocol like websockets.

Roughly:

```js
websocket.disconnected(() => subscription.return())
try {
  for await (let result of subscription) {
    websocket.publish(result);
  }
} finally {
  websocket.close();
}
```